driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -175,7 +178,10 @@ static int lengthFor(ProtocolVersion version) {
                 return set;
             }
 
-            static int serialize(EnumSet<Flag> flags) {
+            static int serialize(EnumSet<Flag> flags, ProtocolVersion protocolVersion) {
+                if (protocolVersion.toInt() < ProtocolVersion.V5.toInt() && flags.contains(USE_BETA))
```

Comment:
Actually I tested moved the check to `Frame.BetaVersionFlagAdder` and it turns out that it is not a good idea to throw exceptions from channel handlers as these would defunct the connection and throw a rather cryptic error message ("Error writing").
So instead I would suggest to move the check to `Frame.BetaVersionFlagAdder` but in a more lenient version:

```java
static class BetaVersionFlagAdder extends MessageToMessageEncoder<Frame> {
    @Override
    protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
        ProtocolVersion protocolVersion = frame.header.version;
        if (protocolVersion.toInt() >= ProtocolVersion.V5.toInt())
            frame.header.flags.add(USE_BETA);
        out.add(frame);
    }
}
```

One might wonder why the check is required at all, since this handler is only added to the pipeline if the user requests the `USE_BETA` flag explicitly. It is still required in the case of renegotiation: e.g. the driver tries `V5` + `USE_BETA`, gets a `PROTOCOL_ERROR` because the server only handles `V4`, and then retries with `V4`: in this case, because `USE_BETA` is not supported in `V4`, the handler should check the version and refrain itself from adding the flag.